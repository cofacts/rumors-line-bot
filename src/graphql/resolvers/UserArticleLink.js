export default {
  /**
   * @param {UserArticleLink} source
   */
  article(source) {
    // Returns stub
    return { id: source.articleId };
  },
};
