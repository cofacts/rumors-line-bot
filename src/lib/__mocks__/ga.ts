const eventMock = jest.fn();
const sendMock = jest.fn();
const setMock = jest.fn();

const ga = Object.assign(
  jest.fn().mockImplementation(() => ({
    event: eventMock,
    send: sendMock,
    set: setMock,
    screenview: () => undefined,
  })),
  {
    clearAllMocks: () => {
      eventMock.mockClear();
      sendMock.mockClear();
      setMock.mockClear();
      ga.mockClear();
    },
    eventMock,
    sendMock,
    setMock,
  }
);

export default ga;
