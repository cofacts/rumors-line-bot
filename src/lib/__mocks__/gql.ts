const mockResultQueue: object[] = [];

function gqlMock() {
  return () => {
    const ret = mockResultQueue.shift();
    return Promise.resolve(ret);
  };
}

/**
 * Enqueues mock data what will be returned in next gql()() call.
 *
 * Once gql()() is invoked, the first data in queue will be consumed.
 *
 * @param {*} returnValue the mock value that gql()()'s returned promise will resolve to.
 * @returns nothing
 */
gqlMock.__push = function (returnValue: object) {
  mockResultQueue.push(returnValue);
};

/**
 * Empty the return data queue.
 *
 * Invoking this in afterEach() will ensure 1 failed test would not leave side effects to other
 * tests.
 */
gqlMock.__reset = function () {
  mockResultQueue.length = 0;
};

/**
 * @returns {Boolean} if the mock response has been depleted
 */
gqlMock.__finished = function () {
  return mockResultQueue.length === 0;
};

export default gqlMock;
export type MockedGql = typeof gqlMock;
