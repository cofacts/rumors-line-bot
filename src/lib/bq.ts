/**
 * BigQuery client and schema
 */

import { BigQuery } from '@google-cloud/bigquery';

const bqDataset = new BigQuery().dataset(
  process.env.BIGQUERY_ANALYTICS_DATASET || ''
);

export function insertEventBatch(eventBatch: EventBatch) {
  const table = bqDataset.table(TABLE);
  return table.insert(eventBatch, { schema: bqSchema });
}
