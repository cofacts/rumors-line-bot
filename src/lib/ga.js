import ua from 'universal-analytics';
import { gaTitle } from './sharedUtils';

/**
 * Sends a screen view and returns the visitor
 *
 * @param {string} uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param {string} state - The current state the LINE bot is in
 * @param {string} documentTitle - The article content (user input) currently in search
 * @param {string} messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns {ua.Visitor} universal analytics visitor
 */
export default function ga(
  uuid,
  state = 'N/A',
  documentTitle = '',
  messageSource = 'user'
) {
  const visitor = ua(process.env.GA_ID, uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  return visitor;
}
