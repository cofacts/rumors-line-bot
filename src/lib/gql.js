import fetch from 'node-fetch';
import rollbar from './rollbar';
import url from 'url';

const API_URL =
  process.env.API_URL || 'https://cofacts-api.hacktabl.org/graphql';

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
// GraphQL Protocol: http://dev.apollodata.com/tools/graphql-server/requests.html
//
export default (query, ...substitutions) => (variables, search) => {
  const queryAndVariable = {
    query: String.raw(query, ...substitutions),
  };

  if (variables) queryAndVariable.variables = variables;

  let status;
  const URL = `${API_URL}${url.format({ query: search })}`;

  return fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-app-secret': process.env.APP_SECRET,
    },
    credentials: 'include',
    body: JSON.stringify(queryAndVariable),
  })
    .then(r => {
      status = r.status;
      return r.json();
    })
    .then(resp => {
      if (status === 400) {
        throw new Error(
          `GraphQL Error: ${resp.errors
            .map(({ message }) => message)
            .join('\n')}`
        );
      }
      if (resp.errors) {
        // When status is 200 but have error, just print them out.
        console.error('GraphQL operation contains error:', resp.errors);
        rollbar.error(
          'GraphQL error',
          {
            body: JSON.stringify(queryAndVariable),
            url: URL,
          },
          { resp }
        );
      }
      return resp;
    });
};
