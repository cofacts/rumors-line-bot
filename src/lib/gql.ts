import fetch from 'node-fetch';
import rollbar from './rollbar';
import { format } from 'url';
import Dataloader from 'dataloader';

const API_URL = process.env.API_URL || 'https://dev-api.cofacts.tw/graphql';

type QV = {
  query: string;
  variables?: object;
};

type Resp = { data: object | null; errors: { message: string }[] | undefined };

// Maps URL to dataloader. Cleared after batched request is fired.
// Exported just for unit test.
export const loaders: Record<string, Dataloader<QV, Resp>> = {};

/**
 * Returns a dataloader instance that can send query & variable to the GraphQL endpoint specified by `url`.
 *
 * The dataloader instance is automatically created when not exist for the specified `url`, and is
 * cleared automatically when the batch request fires.
 *
 * @param url - GraphQL endpoint URL
 * @returns A dataloader instance that loads response of the given {query, variable}
 */
function getGraphQLRespLoader(url: string) {
  if (loaders[url]) return loaders[url];

  return (loaders[url] = new Dataloader<QV, Resp>(async (queryAndVariables) => {
    // Clear dataloader so that next batch will get a fresh dataloader
    delete loaders[url];

    // Implements Apollo's transport layer batching
    // https://www.apollographql.com/blog/apollo-client/performance/query-batching/#1bce
    //
    return (
      await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'x-app-secret': process.env.APP_SECRET ?? '',
        },
        body: JSON.stringify(queryAndVariables),
      })
    ).json();
  }));
}

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
export default (query: TemplateStringsArray, ...substitutions: string[]) =>
  <QueryResp extends object, Variable>(
    variables: Variable,
    search?: Record<string, string | number>
  ) => {
    const queryAndVariable: QV = {
      query: String.raw(query, ...substitutions),
    };

    if (variables) queryAndVariable.variables = variables;

    const URL = `${API_URL}${format({ query: search })}`;

    return getGraphQLRespLoader(URL)
      .load(queryAndVariable)
      .then((resp) => {
        // We cannot get status code of individual request in transport layer batching.
        // but we can guess that it's not 2xx if `data` is null or does not exist.
        // Ref: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#status-codes
        //
        if (!('data' in resp) || !resp.data || typeof resp.data !== 'object') {
          const errorStr =
            'errors' in resp && Array.isArray(resp.errors)
              ? resp.errors.map(({ message }) => message).join('\n')
              : 'Unknown error';

          throw new Error(`GraphQL Error: ${errorStr}`);
        }

        if ('errors' in resp && resp.errors) {
          console.error('GraphQL operation contains error:', resp.errors);
          rollbar.error(
            'GraphQL error',
            {
              body: JSON.stringify(queryAndVariable),
              url: URL,
            },
            { resp }
          );
        }

        return resp as GqlResponse<QueryResp>;
      });
  };
