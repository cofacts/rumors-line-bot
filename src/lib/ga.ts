import ua from 'universal-analytics';
import type { EventParams } from 'universal-analytics';
import { gaTitle } from './sharedUtils';
import { insertEventBatch } from './bq';
import type { EventBatch } from 'src/rumors-db/bq/events';
import rollbar from './rollbar';

/**
 * Sends a screen view and returns the visitor
 *
 * @param uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param state - The current state the LINE bot is in
 * @param documentTitle - The article content (user input) currently in search
 * @param messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns universal analytics visitor
 */
export default function ga(
  uuid: string,
  state = 'N/A',
  documentTitle = '',
  messageSource: 'user' | 'group' | 'room' = 'user'
) {
  const visitor = ua(process.env.GA_ID || '', uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', gaTitle(documentTitle));
  }

  visitor.set('cd1', messageSource);

  let events: EventBatch['events'] = [];
  const extra: Record<string, unknown> = {};

  return {
    set(key: string, value: unknown) {
      extra[key] = value;
      return visitor.set(key, value);
    },

    event(evt: EventParams) {
      events.push({
        category: evt.ec ?? null,
        action: evt.ea ?? null,
        label: evt.el ?? null,
        value: evt.ev === undefined ? null : +evt.ev,
        time: new Date(),
      });
      return visitor.event(evt);
    },
    send() {
      insertEventBatch({
        text: documentTitle,
        messageSource,
        events,
        createdAt: new Date(),
        extra,
      }).catch((e) => {
        console.error('[insertAnalytics]', e);
        rollbar.error(`[insertAnalytics] ${e.message}`, e);
      });
      events = [];
      return visitor.send();
    },
  };
}
