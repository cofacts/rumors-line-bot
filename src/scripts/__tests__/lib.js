jest.mock('src/lib/gql');
jest.mock('src/lib/ga');
jest.mock('src/lib/redisClient');
jest.mock('src/webhook/lineClient');
jest.mock('src/lib/sendMessage');

import lib from '../lib';
import SendMessage from 'src/lib/sendMessage';
import UserArticleLink from 'src/database/models/userArticleLink';
import UserSettings from 'src/database/models/userSettings';
import gql from 'src/lib/gql';
import ga from 'src/lib/ga';

beforeEach(async () => {
  ga.clearAllMocks();
  gql.__reset();
});

describe('get notification list', () => {
  it('should get notification list', async () => {
    gql.__push({
      data: {
        ListArticles: {
          pageInfo: {
            firstCursor: 'WzE1OTM1OTAzNjgyOTQsIjN2dmsxMGNlYTgxdzEiXQ==',
            lastCursor: 'WzE1NDUyNjQ5NzEwOTUsIjFoendmbnZnYnYwdWkiXQ==',
          },
          totalCount: 199,
        },
      },
    });

    gql.__push({
      data: {
        ListArticles: {
          edges: [
            {
              node: {
                id: 'a1',
                articleReplies: [
                  {
                    createdAt: '2020-06-14T11:57:25.100Z',
                  },
                  {
                    createdAt: '2020-05-29T11:57:25.100Z',
                  },
                ],
              },
              cursor: 'WzE1OTM0Njk4MjQyOTcsIjFmaTdkbGE5d3R3am4iXQ==',
            },
            {
              node: {
                id: 'a2',
                articleReplies: [
                  {
                    createdAt: '2020-05-29T10:32:34.103Z',
                  },
                ],
              },
              cursor: 'WzE1OTA3NDgzNTQxMDMsIjM2bngyN3Z1enhuZXAiXQ==',
            },
            {
              node: {
                articleReplies: [],
              },
              cursor: 'WzE1OTA3NDc3Mjk4MTIsIjJ3aTlkdTlhMnN3MTMiXQ==',
            },
            {
              // ListArticle(repliedAt) includes an article as long as it has one articleReply within the time range
              node: {
                id: 'a3',
                articleReplies: [
                  {
                    createdAt: '2020-05-28T10:22:09.812Z',
                  },
                  {
                    createdAt: '2020-05-27T10:22:09.812Z',
                  },
                  {
                    createdAt: '2020-01-25T10:22:09.812Z',
                  },
                ],
              },
              cursor: 'WzE1OTA1NzE2MTkwNTcsIkFWMXZwMGw0eUNkUy1uV2h1YzRlIl0=',
            },
            {
              node: {
                id: 'a4',
                articleReplies: [
                  {
                    createdAt: '2020-05-26T10:21:32.848Z',
                  },
                ],
              },
              cursor: 'WzE1OTAwNTYyODY3MTksIjE2YW95dXY2dDBjd3oiXQ==',
            },
            {
              node: {
                id: 'a5',
                articleReplies: [
                  {
                    createdAt: '2020-05-26T10:21:32.848Z',
                  },
                ],
              },
              cursor: 'WzE1NDUyNjQ5NzEwOTUsIjFoendmbnZnYnYwdWkiXQ==',
            },
          ],
        },
      },
    });

    const fixtures = [
      {
        userId: 'u2',
        articleId: 'a1',
        createdAt: new Date('2020-01-01T18:10:18.314Z'),
        lastViewedAt: new Date('2020-01-01T18:10:18.314Z'),
      },
      {
        userId: 'u1',
        articleId: 'a3',
        createdAt: new Date('2020-01-01T21:10:18.314Z'),
        lastViewedAt: new Date('2020-01-01T21:10:18.314Z'),
      },
      {
        userId: 'u1',
        articleId: 'a4',
        createdAt: new Date('2020-01-01T20:10:18.314Z'),
        lastViewedAt: new Date('2020-01-01T20:10:18.314Z'),
      },
      {
        userId: 'u2',
        articleId: 'a3',
        createdAt: new Date('2020-07-01T23:10:18.314Z'),
        // u2 viewed a3's new reply
        lastViewedAt: new Date('2020-07-01T23:10:18.314Z'),
      },
      {
        userId: 'u3',
        articleId: 'a3',
        createdAt: new Date('2020-01-01T22:10:18.314Z'),
        lastViewedAt: new Date('2020-01-01T22:10:18.314Z'),
      },
    ];

    if (await UserArticleLink.collectionExists()) {
      await (await UserArticleLink.client).drop();
    }

    for (const fixture of fixtures) {
      await UserArticleLink.create(fixture);
    }

    const result = await lib.getNotificationList(
      '2020-05-20T00:00:00.000Z',
      '2020-06-15T00:00:00.000Z'
    );

    expect(result).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
  });

  it('handles empty ListArticles data', async () => {
    gql.__push({
      data: {
        ListArticles: {
          pageInfo: {
            firstCursor: null,
            lastCursor: null,
          },
          totalCount: 0,
        },
      },
    });

    const result = await lib.getNotificationList(
      '2020-05-20T00:00:00.000Z',
      '2020-06-15T00:00:00.000Z'
    );

    expect(result).toMatchSnapshot();
    expect(gql.__finished()).toBe(true);
  });
});

describe('send notification', () => {
  SendMessage.multicast = jest.fn();
  SendMessage.notify = jest.fn();
  const OLD_ENV = process.env;
  const fixtures = [
    {
      userId: 'u1',
      allowNewReplyUpdate: true,
      newReplyNotifyToken: 'this_is_token',
    },
    {
      userId: 'u2',
      allowNewReplyUpdate: false,
      newReplyNotifyToken: 'this_is_token',
    },
    {
      userId: 'u3',
      allowNewReplyUpdate: true,
      newReplyNotifyToken: 'this_is_token',
    },
  ];

  const list = {
    u1: ['a3', 'a4'],
    u2: ['a1', 'a3'],
    u3: ['a3'],
  };
  beforeEach(async () => {
    SendMessage.notify.mockClear();
    SendMessage.multicast.mockClear();

    process.env = { ...OLD_ENV };
    delete process.env.NOTIFY_METHOD;

    if (await UserSettings.collectionExists()) {
      await (await UserSettings.client).drop();
    }
    for (const fixture of fixtures) {
      await UserSettings.create(fixture);
    }
  });

  afterEach(() => {
    process.env = OLD_ENV;
  });

  it('should send notification by multicast api', async () => {
    process.env.NOTIFY_METHOD = 'PUSH_MESSAGE';

    await lib.sendNotification(list);
    expect(SendMessage.notify).not.toHaveBeenCalled();
    expect(SendMessage.multicast).toHaveBeenCalledTimes(1);
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('should send notification by LINE Notify', async () => {
    process.env.NOTIFY_METHOD = 'LINE_NOTIFY';

    await lib.sendNotification(list);
    // called twice because u2 doesn't allow to send notification
    expect(SendMessage.notify).toHaveBeenCalledTimes(2);
    expect(SendMessage.multicast).not.toHaveBeenCalled();
    expect(ga.sendMock).toHaveBeenCalledTimes(1);
  });

  it('handles empty list', async () => {
    process.env.NOTIFY_METHOD = 'PUSH_MESSAGE';

    await lib.sendNotification({});
    expect(SendMessage.notify).not.toHaveBeenCalled();
    expect(SendMessage.multicast).not.toHaveBeenCalled();
    expect(ga.sendMock).not.toHaveBeenCalled();
  });
});

it('test getUserArticleLinksBatch', async () => {
  const fixtures = [
    {
      userId: 'u2',
      articleId: 'a1',
      createdAt: new Date('2020-01-01T18:10:18.314Z'),
      lastViewedAt: new Date('2020-01-01T18:10:18.314Z'),
    },
    {
      userId: 'u1',
      articleId: 'a3',
      createdAt: new Date('2020-01-01T21:10:18.314Z'),
      lastViewedAt: new Date('2020-01-01T21:10:18.314Z'),
    },
    {
      userId: 'u2',
      articleId: 'a3',
      createdAt: new Date('2020-07-01T23:10:18.314Z'),
      lastViewedAt: new Date('2020-07-01T23:10:18.314Z'),
    },
  ];

  if (await UserArticleLink.collectionExists()) {
    await (await UserArticleLink.client).drop();
  }

  for (const fixture of fixtures) {
    await UserArticleLink.create(fixture);
  }
  const articleIds = ['a1', 'a2', 'a3', 'a4'];
  const generator = await lib.getUserArticleLinksBatch(articleIds, 2);
  const result1 = await generator.next();
  expect(result1.value.length).toBe(2);
  expect(result1.done).toEqual(false);
  const result2 = await generator.next();
  expect(result2.value.length).toBe(1);
  expect(result2.done).toEqual(false);
  const result3 = await generator.next();
  expect(result3.value).toMatchInlineSnapshot(`undefined`);
  expect(result3.done).toEqual(true);
});
