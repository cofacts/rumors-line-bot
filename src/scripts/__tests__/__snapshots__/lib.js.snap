// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get notification list handles empty ListArticles data 1`] = `Object {}`;

exports[`get notification list should get notification list 1`] = `
Object {
  "u1": Array [
    "a3",
    "a4",
  ],
  "u2": Array [
    "a1",
  ],
  "u3": Array [
    "a3",
  ],
}
`;
